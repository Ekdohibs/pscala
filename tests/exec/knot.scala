
class Omega { def omega() : Nothing = omega() }

class Function[-A,+B] extends Omega {
  def apply(x: A) : B = omega()
}

class State(s: String) extends Function[Int,Unit] {
  var f = new Function[Int,Unit]();
  def knot(g: Function[Int,Unit]) { f = g };
  override def apply(x: Int) : Unit = if(x == 0) print(s) else f.apply(x-1)
}

object Main {
  val even = new State("even\n");
  val odd = new State("odd\n");
  def knot() { even.knot(odd); odd.knot(even) };
  def main(a: Array[String]) {
    knot();
    var i = 42;
    while(i >= 0) { print(i); print(" "); even.apply(i); i = i - 1 }
  }
}

