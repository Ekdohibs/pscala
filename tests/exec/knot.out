42 even
41 odd
40 even
39 odd
38 even
37 odd
36 even
35 odd
34 even
33 odd
32 even
31 odd
30 even
29 odd
28 even
27 odd
26 even
25 odd
24 even
23 odd
22 even
21 odd
20 even
19 odd
18 even
17 odd
16 even
15 odd
14 even
13 odd
12 even
11 odd
10 even
9 odd
8 even
7 odd
6 even
5 odd
4 even
3 odd
2 even
1 odd
0 even
